### ISSUE TITLE
YOLO object detection: fine-tune for beamline samples

### ISSUE THREAD

**Issue by bondada-a** [bondada-a/erobs#42]
## Summary

YOLO-based object detection has been integrated into the beambot MCP server (`detect_objects(method="yolo")`). A pre-trained model on the ChemistryLabApparatus-25 dataset (25 classes, 98.3% mAP@50) is available at `~/.beambot/models/beambot_chemlab_best.pt`. This issue tracks the remaining work to make it production-ready for beamline-specific objects.

## What's Built

- [x] `yolo_detector.py` — YoloDetector class with GPU inference, model caching, annotation
- [x] MCP integration — `detect_objects(method="yolo")` with model/confidence/class filter params
- [x] Training pipeline — `~/.beambot/training/train_yolo.py` for downloading datasets and fine-tuning
- [x] Pre-trained model — ChemistryLabApparatus-25 (beaker, pipette, test tube, flask, etc.)
- [x] RTX 4060 GPU inference verified (~2ms per image)

### High Priority
- [ ] **Collect beamline-specific training images** — photograph actual samples, tip racks, well plates, vials, and other beamline objects with the Zivid camera from sample_scan positions. ~100-200 images needed.
- [ ] **Label with Roboflow** — upload images, draw bounding boxes, define custom classes (e.g., `sample`, `tip_rack`, `vial`, `well_plate`, `puck`)
- [ ] **Fine-tune on beamline data** — merge ChemistryLab dataset with custom images, train with `train_yolo.py`
- [ ] **Integrate YOLO into vision_stages.py** — add `detection_type: "yolo"` for `vision_moveto` and `vision_pick_place` task types (currently YOLO is MCP-only)

### Medium Priority
- [ ] **Test with YOLOv8s (small)** — larger model may improve accuracy, still fast on RTX 4060
- [ ] **Add YOLO to vision_scan** — batch detection across multiple scan positions
- [ ] **Model versioning** — track which model is deployed, auto-download from a shared location

### Low Priority / Future
- [ ] **YOLOv8-seg** — instance segmentation for precise object boundaries (useful for collision avoidance)
- [ ] **Continuous learning** — save misdetections, periodically retrain
- [ ] **Multi-camera** — run YOLO on ZED images for wide-FOV scene understanding

## Available Datasets

| Dataset | Classes | Images | Notes |
|---------|---------|--------|-------|
| ChemistryLabApparatus-25 (Roboflow) | 25 | 4,600 | **Already trained** — pipette, flask, beaker, etc. |
| Chemistry Lab Object Detection (Roboflow) | ~14 | 4,512 | Similar classes, different images |
| Custom beamline data | TBD | 0 | **Needs collection and labeling** |

## Training Command

```bash
# Re-train on existing dataset
python3 ~/.beambot/training/train_yolo.py \
  --data ~/.beambot/training/datasets/chemlab_apparatus/data.yaml \
  --epochs 50 --name beambot_chemlab

# Train on custom Roboflow project
python3 ~/.beambot/training/train_yolo.py \
  --roboflow-workspace YOUR_WORKSPACE \
  --roboflow-project YOUR_PROJECT \
  --roboflow-api-key YOUR_KEY \
  --epochs 50 --name beambot_custom
```

## Usage

```
# MCP tool
detect_objects(method="yolo")
detect_objects(method="yolo", yolo_model="~/.beambot/models/beambot_chemlab_best.pt")
detect_objects(method="yolo", yolo_classes="Pipette,Beaker", yolo_confidence=0.3)
```
